### PR TITLE
Image Stacking

### PR DESCRIPTION
**Pull request by wpferguson**
Created image_stack.lua based on Pat David's blog entry, https://patdavid.net/2013/05/noise-removal-in-photos-with-median_6.html.  Updated README.md to reflect the new script.

This script adds a new export target, image_stack.  A stack of images are selected, exported, optionally aligned, then processed using the specified operator.  Mean and median are the two most common, but others are supplied.  The resulting image is imported back into the collection.  

An option exists to tag the source images used to create the stacked result.  Images tagged in this manner can be collected using collect images and then used again to create another result.

This script has been tested on Linux, MacOS, and Windows.

I tested this using the two highest ISO's my camera has, 6400 and 12800.  I shot 10 images at ISO 6400 and 12 at ISO 12800.  I first tested just stacking with no preprocessing and got a 2 - 3 stop improvement in noise.  Some preprocessing (demosaic - amaze w/color smoothing 2, profiled denoise wavelets chroma, profiled denoise wavelets lightness) resulted in pretty much noise free images.  I used both mean and median.  Mean was quick, usually less than 10 seconds to process a stack.  Median took between 1 and 2 minutes to process the same stack.  I tried some other operators and their processing time fell between mean and median.

### CODE REVIEW

**Comment by wpferguson** [comment]
@supertobi - Thoughts and/or suggestions?

File: contrib/image_stack.lua [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+    local result = dtsys.external_command(convert_command)
+    if result == 0 then
+      dt.print(_("image stack processed"))
+      cleanup(img_list)
+      job.percent = job.percent + percent_step
+
+      -- import image
+      dt.print(_("importing result"))
+      local film_roll_path = extract_collection_path(image_table)
+      local import_filename = df.create_unique_filename(film_roll_path .. "/" .. df.get_filename(output_filename))
```
Comment (supertobi) [inline contrib/image_stack.lua]: Does this work on Windows?

File: contrib/image_stack.lua [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+      df.file_move(output_filename, import_filename)
+      imported_image = dt.database.import(import_filename)
+      local created_tag = dt.tags.create("Created with|image_stack")
+      dt.tags.attach(created_tag, imported_image)
+      for img,_ in pairs(image_table) do
+        copy_image_attributes(img, imported_image)
+        break
```
Comment (supertobi) [inline contrib/image_stack.lua]: I'm not sure this break is OK or not. But I think so.

File: contrib/image_stack.lua [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+      end
+      job.percent = job.percent + percent_step
+
+      -- tag images if requested
+
+      if tag_source then
+        dt.print(_("tagging source images"))
+        local source_tag = dt.tags.create("Source file|" .. imported_image.filename)
```
Comment (supertobi) [inline contrib/image_stack.lua]: Should this be translatable?

File: contrib/image_stack.lua [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+-- Normally this shouldn't be a problem, but if you use this code just beware.
+-- If you want to do it non recursively, then remove the /s argument from dir
+-- and grab the path component from the search string and prepend it to the files
+-- found.
+
+local function list_files(search_string)
+  local ls = "ls "
+  local files = {}
+  local dir_path = nil
+  local cnt = 1
```
[the remark this answers is not in the document]
Comment (wpferguson) [reply contrib/image_stack.lua]: Fortran?  All my command line loops start with for i in ...  :-)

File: contrib/image_stack.lua (hunk 1 above)
Comment (wpferguson) [reply contrib/image_stack.lua]: It actually does as long as you don't send it to the shell.  But, I should probably do the PS  substitution just to be safe.

File: contrib/image_stack.lua (hunk 2 above)
Comment (wpferguson) [reply contrib/image_stack.lua]: I'm just copying the attributes from the first image, since they should be the same for all images except for time

Comment (wpferguson) [reply contrib/image_stack.lua]: I'll add a comment

File: contrib/image_stack.lua (hunk 3 above)
Comment (wpferguson) [reply contrib/image_stack.lua]: Yes

**Comment by supertobi** [comment]
Looks good. Feel free to push it.

**Pull request merged**